File: CovarianzaYcontraVarianza_C/Joven.cc
#include "Joven.h"
#include <iostream>
#include<string>
using namespace std;

Joven::Joven(string n,Movilidad mov):Persona(n){
  m  = mov;
}

Joven::Joven(string n,Bicicleta bic):Persona(n){
  b  = bic;
}

Joven::Joven(string n,Scooter sct):Persona(n){
  s  = sct;
}

void Joven::avanzar(){
  cout<<"El joven "<<nombre<<" avanza"<<endl;
}

void Joven::mostrarCovarianza(){
  cout<<"COVARIANZA"<<endl;
  //s.avanzar();
}

File: Covarianza y contravarianza java/src/to/Joven.java
package to;

public class Joven extends Persona{
	private Movilidad movilidad;
	private Bicicleta bicicleta;
	private Scooter scooter;
	
	Joven(String nombre, Movilidad movilidad) {
		super(nombre);
		this.movilidad = movilidad;
	}
	
	Joven(String nombre, Scooter scooter) {
		super(nombre);
		this.scooter = scooter;
	}
	
	Joven(String nombre, Bicicleta bicicleta) {
		super(nombre);
		this.bicicleta = bicicleta;
	}
	
	public void mostrarContraCovarianza() {
		System.out.println("CONTRACOVARIANZA");
		System.out.print(nombre +" usa la " + movilidad.toString() + "asi que: ");
		movilidad.avanzar();
	}
	
	public void mostrarCovarianza() {
		System.out.println("COVARIANZA");
		System.out.print(nombre +" usa un(a) " + scooter.toString() + " asi que: ");
		scooter.avanzar();
	}
	public void mostrarCovarianza2() {
		System.out.println("COVARIANZA");
		System.out.print(nombre +" usa un(a) " + bicicleta.toString() + " asi que: ");
		bicicleta.avanzar();
	}
}
